src/types.ts

```diff
@@ -24,12 +24,69 @@ export interface Tags {
   };
 }
 
-export interface TokenList {
+// Parameters for investing into a protected investment.
+//
+// The function signature is constant, which is why it's an optional property. The `tokens` property defines the
+// set of input tokens that can be used to deposit into this investment opportunity. The items in the function
+// signature correspond to these `tokens`. For example, if we're investing in the 3Crypto pool, the `tokens` array
+// would be [ETH, WBTC, USDT]. When calling `invest` on `address`, amounts[0] would correspond to the amount of ETH
+// to invest, amounts[1] would correspond to the amount of WBTC to invest, etc. An amount of MAX_UINT256 should be
+// used to signal that you want to invest all tokens you have
+//
+// The same concept also applies to `approves`, which defines whether or not to check and execute approvals for the
+// given token. This is an option to save gas. Similar to `amounts`, the number passed in here specifies how much
+// to approve the spending contract for
+export interface InvestScript {
+  readonly address: string;
+  readonly tokens: {
+    readonly input: TokenInfo[]; // tokens required to invest in this opportunity
+    readonly receipt: TokenInfo[]; // receipt tokens received for investing
+    readonly rewards: TokenInfo[]; // rewards tokens received for investing
+  };
+  readonly signature?: 'function invest(uint256[] amounts, uint256[] approve)';
+}
+
+// Parameters for exiting a protected investment.
+//
+// Function signature meaning is analogous to `InvestScript`. The `amounts` and `claimRewards` parameters are uint256
+// where zero means don't withdraw/claim anything and MAX_UINT256 means withdraw/claim all. Note that not all
+// reward tokens will support arbitrary claim amounts. The `recipient` field specifies the address all output
+// tokens and reward tokens should be sent to
+export interface DivestScript {
+  readonly address: string;
+  readonly tokens: {
+    readonly receipt: TokenInfo[]; // receipt tokens the user needs to exit this opportunity
+    readonly output: TokenInfo[]; // tokens the user will receive after exiting this opportunity
+    readonly rewards: TokenInfo[]; // rewards tokens the user may claim when exiting this opportunity
+  };
+  readonly signature?: 'function divest(uint256[] amounts, address recipient, uint256[] claimRewards)';
+}
+
+export interface Data {
+  // TODO
```

Comment:
Right, so I think the only thing we need here are (and I think it may be worth moving up to `ProtectionMarketInfo` since it's just two fields:

- The HTTP endpoint to fetch current market data and position
- Some structure describing the legs of the strategy, e.g. for the saddle aleth investment opportunity it's:

```
      strategy: [
        { action: 'Borrow', destination: 'Ether', imagePath: '/tokens/eth.png' },
        { action: 'Deposit to', destination: 'Saddle', imagePath: '/platforms/saddle.png' },
        { action: 'Stake on', destination: 'Alchemix', imagePath: '/platforms/alchemix.png' },
      ],
```

although we can probably remove the `imagePath`s and turn the actions/destinations into enums

Reply:
Image path seems like a good field, maybe we make it optional. Can you and @dpekar help me come up with a good list of types for the `action` enum? Feel free to just share it here and I can update the type definition

Reply:
I think @tonysheng  is actually the right person to come up with a list of the action types.

Made a ticket for him here: 
https://linear.app/cozy-finance/issue/ENG-358/come-up-with-list-of-types-of-actions-leg-of-investments-can-take

Reply:
Also, should we use the ids from `platformIds` (that are part of protection markets) for the destination field here?

Seems like a good idea to reuse the identifiers to make it easy for the front end.  I think imagePath being optional is good. When it's null, frontends can fall back to using the images for the ids in the  destination.

Could also add a field destinationPlatformId or something if we want to keep the destination field readable.

Reply:
I agree the current shape doesn't make perfect sense this way.

We basically have:
1. static data about the market
2. static data about the associated invest opty
3.  dynamic data about the invest opty

3 is handled by the endpoint, so we don't really need a data field.

So more like, 
```
interface ProtectionMarketInfo {
  readonly underlying: TokenInfo; // if `underlying.address === 0xEeee....EEeE`, underlying is ETH
  readonly trigger: string; // trigger contract address for protection markets, or zero address for money markets
  readonly isTriggered: boolean; // true if this market was triggered

  investmentOpportunity?: {
    readonly invest: InvestScript ;
    readonly divest: DivestScript;  
    readonly dataURI: string;  
    readonly strategy: StrategyLeg[];
 }
}
```

Reply:
All of this sounds good to me, I like the platform IDs idea. I think we could actually use this repo as the master copy that defines the mapping from platform ID to platform name. 

Also I like the proposed `ProtectionMarketInfo` format. Do we need to break `strategy` into an invest and divest portion? 

Do you think all of the `investmentOpportunity` fields should be required? I'm on the fence about this one

Reply:
> Also I like the proposed ProtectionMarketInfo format. Do we need to break strategy into an invest and divest portion?
Nope!

> Do you think all of the investmentOpportunity fields should be required? I'm on the fence about this one
I think the min set of fields for the opty to be functional should be required and anything else is optional.  In this case, that means invest, divest & dataURI required and strategy optional.